src/serialisation.rs

```diff
@@ -34,15 +35,13 @@ pub fn read_n(n: usize, bytes: &[u8]) -> Result<(&[u8], &[u8]), Error> {
 /// Reads 32 bytes and converts it to a BlsScalar
 /// Returns the remaining bytes
 pub fn read_scalar(bytes: &[u8]) -> Result<(BlsScalar, &[u8]), Error> {
-    let (bytes32, rest) = read_n(32, bytes)?;
-    let mut arr32 = [0u8; 32];
-    arr32.copy_from_slice(bytes32);
-    let scalar = BlsScalar::from_bytes(&arr32);
-    if scalar.is_none().into() {
-        return Err(SerialisationErrors::BlsScalarMalformed.into());
-    }
-    Ok((scalar.unwrap(), rest))
+    let (bytes, rest) = read_n(BlsScalar::SIZE, bytes)?;
+
+    BlsScalar::from_slice(bytes)
```

Comment:
I'm okay to keep as is for the time being but this is not optimal, since we're checking for the length twice, in `read_n` and in `from_slice`.
Probably the best approach would be remove `read_n` from the `read_*` where we're using `from_slice` and just mapped `BadLength` to `NotEnoughBytes` and `InvalidData` to the current type error (e.g. `SerialisationErrors::BlsScalarMalformed`) (if we don't want to change too much the API, I mean).

Reply:
I agree with your concern, and this specific use-case is interesting to add as a feature to `dusk-bytes`.

In this usage, we need to not only parse the targeted type, but also split the slice into the tuple composed by the used bytes and the remainder ones.

This pattern is broadly used in crates such as `nom` and will be very helpful to us!

Reply:
https://github.com/dusk-network/dusk-bytes/issues/11